nav2_behaviors/src/spin.cpp

```diff
@@ -0,0 +1,198 @@
+
+  RCLCPP_INFO(get_logger(), "Initialized the Spin  behavior");
+}
+
+Spin::~Spin()
+{
+  RCLCPP_INFO(get_logger(), "Shutting down the Spin behavior");
+}
+
+nav2_tasks::TaskStatus Spin::spin(const nav2_tasks::SpinCommand::SharedPtr command)
+{
+  RCLCPP_INFO(get_logger(), "Attempting to spin");
+
+  double yaw, pitch, roll;
+  getAnglesFromQuaternion(command->quaternion, yaw, pitch, roll);
+
+  if (roll != 0.0 || pitch != 0.0) {
+    RCLCPP_INFO(get_logger(), "Spinning on Y and X not supported, "
+      "will only spin in Z.");
+  }
+
+  RCLCPP_INFO(get_logger(), "Currently only supported spinning by a fixed amount");
+  TaskStatus status = timedSpin();
+
+  nav2_tasks::SpinResult result;
+  task_server_->setResult(result);
+
+  return status;
+}
+
+nav2_tasks::TaskStatus Spin::timedSpin()
+{
+  auto start_time = std::chrono::system_clock::now();
+  auto time_since_msg = std::chrono::system_clock::now();
+
+  while (true) {
+    if (task_server_->cancelRequested()) {
+      RCLCPP_INFO(get_logger(), "Task cancelled");
+      task_server_->setCanceled();
+      return TaskStatus::CANCELED;
+    }
+
+    // TODO(orduno): For now, spinning is time based
+    auto duration = 4s;
+
+    // Log a message every second
+    auto current_time = std::chrono::system_clock::now();
+    if (current_time - time_since_msg >= 500ms) {
+      RCLCPP_INFO(get_logger(), "Spinning...");
+      time_since_msg = std::chrono::system_clock::now();
+    }
+
+    // Output control command
+    geometry_msgs::msg::Twist cmd_vel;
+    cmd_vel.linear.x = 0.0;
+    cmd_vel.linear.y = 0.0;
+    cmd_vel.angular.z = 0.5;
+    vel_pub_->publish(cmd_vel);
+
+    if (current_time - start_time >= duration) {
+      cmd_vel.angular.z = 0.0;
+      vel_pub_->publish(cmd_vel);
+      RCLCPP_INFO(get_logger(), "Completed rotation");
+      break;
+    }
+  }
+  return TaskStatus::SUCCEEDED;
+}
+
+nav2_tasks::TaskStatus Spin::controlledSpin()
+{
+  // TODO(orduno) Test controller
+
+  // Get current robot orientation
+  double start_yaw, current_yaw;
+
+  if (!getRobotYaw(start_yaw)) {
+    return TaskStatus::FAILED;
+  }
+
+  while (true) {
+    if (task_server_->cancelRequested()) {
+      RCLCPP_INFO(get_logger(), "Task cancelled");
+      task_server_->setCanceled();
+      return TaskStatus::CANCELED;
+    }
+
+    getRobotYaw(current_yaw);
+
+    double current_angle = current_yaw - start_yaw;
+
+    double dist_left = M_PI - current_angle;
+
+    // TODO(orduno) forward simulation to check if future position is feasible
+
+    // compute the velocity that will let us stop by the time we reach the goal
+    // v_f^2 == v_i^2 + 2 * a * d
+    // solving for v_i if v_f = 0
+    double vel = sqrt(2 * rotational_acc_lim_ * dist_left);
+
+    // limit velocity
+    vel = std::min(std::max(vel, min_rotational_vel_), max_rotational_vel_);
+
+    geometry_msgs::msg::Twist cmd_vel;
+    cmd_vel.linear.x = 0.0;
+    cmd_vel.linear.y = 0.0;
+    cmd_vel.angular.z = vel;
```

Comment:
This rotational velocity should probably scale with angular distance so that it doesnt abruptly start and stop or cause massive wheel slippage when accelerating

Reply:
Some scaling is done in 153 considering the acceleration limit and distance left.